### PR TITLE
Data Export - Create Job

### PR DESCRIPTION
**Pull request by ChrisHuynh333**
## What does this PR do and why?
This PR adds the Data Export `CreateJob`. Because exports have the potential to be very large files, we first initiate the export creation in the `CreateService`, which then calls the `CreateJob` to create and attach the actual export file in the background.

Note: Email notification will be handled in a [separate PR](https://github.com/phac-nml/irida-next/pull/464)
Note: Analysis exports will be added when analysis storage is implemented. 

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

## How to set up and validate locally

- Open `rails c`
- Reseed your database if needed `bin/rails db:drop/create/migrate/seed`

- Set up your params:
```
user = User.find_by(email: 'user1@email.com')
params_1 = {'export_type' => 'sample', 'export_parameters' => {'ids' => [Sample.find_by(name: 'Streptococcus agalactiae/Outbreak 2022 Sample 10').id, Sample.find_by(name: 'Streptococcus agalactiae/Outbreak 2022 Sample 9').id, Sample.find_by(name: 'Neisseria gonorrhoeae/Outbreak 2022 Sample 10').id]}, 'email_notification' => true, 'name' => 'test export'}
```
- Navigate to `Streptococcus agalactiae/Outbreak 2022 Sample 10` sample page, and upload:
  - A set of paired end attachments (can be found in `test/fixtures/files`). An example of files would be `08-5578-small_S1_L001_R1_001.fastq` and `08-5578-small_S1_L001_R2_001.fastq`
  - A unknown file format (not a `fasta` or `fastq` file)

- Create an export with: `DataExports::CreateService.new(user, params_1).execute`
- Set: `data_export = user.data_exports.last`
- After 30 seconds, ensure `data_export.status = ready` and `data_export.expires_at` is 3 business days from the current date

Download the zip file
- To download the zip file, in the console run: `Rails.application.routes.url_helpers.rails_blob_url(data_export.file)`
- Copy and paste the above output into a browser. This should download the zip.
- Check Zip File:
  - Your file directories should be the following:
```
|---- manifest.json
|---- S. agalactiae outbreak 2022 project PUID
|                          |--------- Sample 10 PUID
|                          |                   | ---- test_file_1.fastq PUID
|                          |                   |                      |------- test_file_1.fastq
|                          |                   |   
|                          |                   | ---- unknown file PUID
|                          |                   |                      |------- unknown file.format
|                          |                   |   
|                          |                   | ---- paired files PUID
|                          |                                          |------- paired file 1
|                          |                                          |------- paired file 2
|                          |  
|                          |--------- Sample 9 PUID
|                                              | ---- test_file_1.fastq PUID
|                                                                     |------- test_file_1.fastq
|---- N. gonorrhoeae/Outbreak 2022 project PUID
                           |--------- Sample 10 PUID
                                               | ---- test_file_1.fastq PUID
                                                                     |------- test_file_1.fastq
```

  - Check the `manifest.json` is the following:
```
{
  "type": "Samples Export",
  "date": "Date.current",
  "children": [
    {
      "name": "PUID S. agalactiae Outbreak 2022",
      "type": "folder",
      "irida-next-type": "project",
      "irida-next-name": "Outbreak 2022",
      "children": [
        {
          "name": "PUID Streptococcus agalactiae/Outbreak 2022 Sample 10",
          "type": "folder",
          "irida-next-type": "sample",
          "irida-next-name": "Streptococcus agalactiae/Outbreak 2022 Sample 10",
          "children": [
            {
              "name": "PUID Paired End files",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "08-5578-small_S1_L001_R1_001.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq",
                    "direction": "forward",
                    "type": "illumina_pe"
                  }
                },
                {
                  "name": "08-5578-small_S1_L001_R2_001.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq",
                    "direction": "reverse",
                    "type": "illumina_pe"
                  }
                }
              ]
            },
            {
              "name": "PUID unknown file",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "Screenshot from 2024-03-11 11-42-46.png",
                  "type": "file",
                  "metadata": {
                    "format": "unknown"
                  }
                }
              ]
            },
            {
              "name": "PUID test_file_1.fastq",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "test_file_1.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq"
                  }
                }
              ]
            }
          ]
        },
        {
          "name": "PUID Streptococcus agalactiae/Outbreak 2022 Sample 9",
          "type": "folder",
          "irida-next-type": "sample",
          "irida-next-name": "Streptococcus agalactiae/Outbreak 2022 Sample 9",
          "children": [
            {
              "name": "PUID test_file_1.fastq",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "test_file_1.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq"
                  }
                }
              ]
            }
          ]
        }
      ]
    },
    {
      "name": "PUID Neisseria gonorrhoeae Outbreak 2022 Project",
      "type": "folder",
      "irida-next-type": "project",
      "irida-next-name": "Outbreak 2022",
      "children": [
        {
          "name": "PUID Neisseria gonorrhoeae/Outbreak 2022 Sample 10",
          "type": "folder",
          "irida-next-type": "sample",
          "irida-next-name": "Neisseria gonorrhoeae/Outbreak 2022 Sample 10",
          "children": [
            {
              "name": "PUID test_file_1.fastq",
              "type": "folder",
              "irida-next-type": "attachment",
              "children": [
                {
                  "name": "test_file_1.fastq",
                  "type": "file",
                  "metadata": {
                    "format": "fastq"
                  }
                }
              ]
            }
          ]
        }
      ]
```
## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/jobs/data_exports/create_job.rb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    def initialize_manifest(export_type)
+      @manifest = if export_type == 'sample'
+                    { 'type' => 'Samples Export', 'date' => Date.current, 'children' => [] }
+                  else
+                    { 'type' => 'Analysis Export', 'date' => Date.current, 'children' => [] }
+                  end
+    end
+
+    def create_sample_zip(data_export) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
```
Comment (deepsidhu85) [inline app/jobs/data_exports/create_job.rb:29]: This method seems to have become unnecessarily large and requires a few rubocop disables.

This method can be split up to remove all the rubocop disables by adding 2 new methods (write_attachments, and write_manifest) and moving the relevant code into the methods

- write_attachments should take params (project, sample, zip)
- write_manifest should take params (manifest, zip)

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Just a quick comment to start to get rid of the rubocop disables

**Review by JeffreyThiessen — CHANGES_REQUESTED** [review]
Overall looks good. While testing I rand into an issue where the job would not run after being queue'd but I don't know if that is the code problem or something wrong on my machine.

I manually started the job with `DataExports::CreateJob.perform_now(data_export)` and it worked as expected.

My main point of feedback is the date/time stuff should be in its own helper file.

File: app/jobs/data_exports/create_job.rb [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      manifest_file.unlink
+    end
+
+    def attach_zip(data_export, zip)
+      data_export.file.attach(io: zip.open, filename: "#{data_export.id}.zip")
+      zip.close
+      zip.unlink
+    end
+
+    def set_expiry
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/jobs/data_exports/create_job.rb]: Fixed in 943a166d78713cba9304c5ce616e297fdfea44ee

File: app/jobs/data_exports/create_job.rb [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+      check_informal_holidays = Holidays.between(Date.current, expiry, %i[ca_bc ca_on ca], :informal)
+
+      extra_days = 0
+      extra_days += add_holidays(check_formal_holidays, observed_holidays) if check_formal_holidays.count.positive?
+      extra_days += add_holidays(check_informal_holidays, informal_holidays) if check_informal_holidays.count.positive?
+      expiry = extra_days.business_days.after(expiry) if extra_days.positive?
+      expiry
+    end
+
+    def add_holidays(days_to_check, holidays)
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/jobs/data_exports/create_job.rb]: Fixed in 943a166d78713cba9304c5ce616e297fdfea44ee

**Review by JeffreyThiessen — APPROVED** [review]
Looks good

File: app/jobs/data_exports/create_job.rb (hunk 1 above)
Comment (deepsidhu85) [reply app/jobs/data_exports/create_job.rb]: Thanks!

**Review by deepsidhu85 — APPROVED** [review]
Looks good to me! Working as expected!

**Pull request merged**